extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java

```diff
@@ -0,0 +1,333 @@
+  }
+
+  @Override
+  public SplittableInputSource<URI> withSplit(InputSplit<URI> split)
+  {
+    return new S3InputSource(s3Client, ImmutableList.of(split.get()), ImmutableList.of());
+  }
+
+  @Override
+  public boolean needsFormat()
+  {
+    return true;
+  }
+
+  @Override
+  protected InputSourceReader formattableReader(
+      InputRowSchema inputRowSchema,
+      InputFormat inputFormat,
+      @Nullable File temporaryDirectory
+  )
+  {
+    return new InputEntityIteratingReader(
+        inputRowSchema,
+        inputFormat,
+        // formattableReader() is supposed to be called in each task that actually creates segments.
+        // The task should already have only one split in parallel indexing,
+        // while there's no need to make splits using splitHintSpec in sequential indexing.
+        createSplits(inputFormat, null).map(split -> new S3Entity(s3Client, split.get())),
+        temporaryDirectory
+    );
+  }
+
+
+  /**
+   * Create an {@link URI} from the given {@link S3ObjectSummary}. The result URI is composed as below.
+   *
+   * <pre>
+   * {@code s3://{BUCKET_NAME}/{OBJECT_KEY}}
+   * </pre>
+   */
+  public static URI toUri(S3ObjectSummary object)
+  {
+    final String originalAuthority = object.getBucketName();
+    final String originalPath = object.getKey();
+    final String authority = originalAuthority.endsWith("/") ?
+                             originalAuthority.substring(0, originalAuthority.length() - 1) :
+                             originalAuthority;
+    final String path = originalPath.startsWith("/") ? originalPath.substring(1) : originalPath;
+
+    return URI.create(StringUtils.format("s3://%s/%s", authority, path));
+  }
+
+  @Override
+  public boolean equals(Object o)
+  {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    S3InputSource that = (S3InputSource) o;
+    return Objects.equals(uris, that.uris) &&
+           Objects.equals(prefixes, that.prefixes);
+  }
+
+  @Override
+  public int hashCode()
+  {
+    return Objects.hash(uris, prefixes);
+  }
+
+  @Override
+  public String toString()
+  {
+    return "S3InputSource{" +
+           "uris=" + uris +
+           ", prefixes=" + prefixes +
+           '}';
+  }
+
+  private Iterable<InputSplit<URI>> getIterableObjectsFromPrefixes()
+  {
+    return () -> objectFetchingIterator(s3Client, prefixes.iterator());
+  }
+
+  /**
+   * iterator which fetches batches of {@link #MAX_LISTING_LENGTH} objects given a set of prefixes, using
+   * {@link ServerSideEncryptingAmazonS3#listObjectsV2}, with a fallback to
+   * {@link ServerSideEncryptingAmazonS3#getObjectMetadata} to check if the 'prefix' is an object in the event the
+   * list objects call responds with a 403 http status code
+   */
+  private static Iterator<InputSplit<URI>> objectFetchingIterator(
```

Comment:
This code looks like it could be shareable between the input source and the firehose. If true, please accomplish that sharing, ideally by having the firehose call into the input source.

Reply:
It is, i was just ignoring the firehose.

I have moved this iterator to `S3Utils` and changed the signature to be `Iterator<S3ObjectSummary>`, and modified `S3Utils.objectSummaryIterator`, used by `StaticS3FirehoseFactory` and `S3TimestampVersionedDataFinder`, to take a `URI` instead of a bucket and key (mildly unfortunate since we will be converting it back to bucket and key, but all the callers have URI available), and defer the logic to this newer iterator.